File: crates/tauri-plugin-ani-player/ios/Sources/AniPlayerPlugin.swift
import Foundation
import SwiftUI
import Tauri
import UIKit

private struct DispatchArgs: Decodable {
    let commandJson: String
}

private enum AniPlayerPluginError: LocalizedError {
    case invalidCommand(String)
    case unavailable(String)

    var errorDescription: String? {
        switch self {
        case .invalidCommand(let message), .unavailable(let message): message
        }
    }
}

/** 将 Rust PlayerTransport 连接到 iOS MobileVLCKit SwiftUI 播放页。 */
final class AniPlayerPlugin: Plugin, UIAdaptivePresentationControllerDelegate {
    private var controller: MobileVLCPlayerController?
    private var hostingController: UIHostingController<PlayerScreen>?
    private var activeSessionID: String?
    private var activeSource: [String: Any]?
    private var sequence: UInt64 = 0
    private var backgroundObserver: NSObjectProtocol?
    private var foregroundObserver: NSObjectProtocol?

    /** 注册应用前后台通知并保持同一原生播放控制器。 */
    override init() {
        super.init()
        backgroundObserver = NotificationCenter.default.addObserver(
            forName: UIApplication.didEnterBackgroundNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.controller?.enterBackground()
        }
        foregroundObserver = NotificationCenter.default.addObserver(
            forName: UIApplication.willEnterForegroundNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.controller?.becomeActive()
        }
    }

    /** 插件释放时移除通知并停止仍在运行的媒体。 */
    deinit {
        if let backgroundObserver { NotificationCenter.default.removeObserver(backgroundObserver) }
        if let foregroundObserver { NotificationCenter.default.removeObserver(foregroundObserver) }
        let playerController = controller
        let playerHost = hostingController
        DispatchQueue.main.async {
            playerController?.close()
            playerHost?.dismiss(animated: false)
        }
    }

    /** 返回 iOS 内置 MobileVLCKit 的稳定能力集合。 */
    @objc public func capabilities(_ invoke: Invoke) {
        invoke.resolve(["capabilitiesJson": Self.encode(Self.capabilitiesPayload())])
    }

    /** 执行 Rust 已校验并解析真实路径的播放器命令。 */
    @objc public func dispatch(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(DispatchArgs.self)
        guard
            let data = args.commandJson.data(using: .utf8),
            let command = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            throw AniPlayerPluginError.invalidCommand("播放器命令不是有效 JSON")
        }
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            do {
                let result = try self.execute(command)
                invoke.resolve(["resultJson": Self.encode(result)])
            } catch {
                NSLog("AniPlayerPlugin command failed: %@", error.localizedDescription)
                invoke.reject(
                    "iOS 播放器命令失败",
                    code: "player_command_failed",
                    error: error
                )
            }
        }
    }

    /** 返回当前原生播放器快照；页面尚未打开时返回空。 */
    @objc public func snapshot(_ invoke: Invoke) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            let snapshotJson = self.snapshotPayload().map(Self.encode)
            let value: Any = snapshotJson ?? NSNull()
            invoke.resolve(["snapshotJson": value])
        }
    }

    /** 幂等关闭原生播放页并清除受控会话。 */
    @objc public func shutdown(_ invoke: Invoke) {
        DispatchQueue.main.async { [weak self] in
            self?.closePlayer(animated: false)
            invoke.resolve(["stopped": true])
        }
    }

    /** 用户通过系统交互关闭页面时同步释放 MobileVLCKit。 */
    func presentationControllerDidDismiss(_ presentationController: UIPresentationController) {
        closePlayer(animated: false)
    }

    /** 将动作映射到当前 MobileVLCKit 控制器。 */
    private func execute(_ command: [String: Any]) throws -> [String: Any] {
        let commandID = try Self.requiredString(command, key: "commandId")
        let sessionID = try Self.requiredString(command, key: "sessionId")
        let action = try Self.requiredString(command, key: "type")
        if action == "load" {
            try launch(command, sessionID: sessionID)
            return Self.accepted(commandID)
        }
        guard activeSessionID == sessionID, let controller else {
            return Self.rejected(commandID, code: "resource-unavailable", message: "播放器会话已切换")
        }

        switch action {
        case "play": controller.play()
        case "pause": controller.pause()
        case "seek":
            controller.seek(to: Self.milliseconds(command["positionSeconds"] as? Double))
        case "set-volume":
            let volume = Int32(((command["volume"] as? Double) ?? 0) * 100)
            controller.setVolume(volume)
        case "set-muted": controller.setMuted((command["muted"] as? Bool) == true)
        case "set-rate": controller.setPlaybackRate(Float((command["rate"] as? Double) ?? 1))
        case "select-audio-track":
            guard let trackID = Int32((command["trackId"] as? String) ?? "") else {
                return Self.rejected(commandID, code: "unsupported", message: "音轨标识无效")
            }
            controller.selectAudioTrack(trackID)
        case "select-subtitle-track":
            let trackID = (command["trackId"] as? String).flatMap { Int32($0) }
            controller.selectSubtitleTrack(trackID)
        case "set-subtitle-scale":
            controller.setSubtitleScale((command["subtitleScale"] as? Int) ?? 100)
        case "set-aspect-ratio":
            guard let ratio = Self.aspectRatio(command) else {
                return Self.rejected(commandID, code: "unsupported", message: "iOS 不支持该画面比例")
            }
            controller.selectAspectRatio(ratio)
        case "set-fullscreen":
            setFullscreen((command["fullscreen"] as? Bool) == true)
        case "set-picture-in-picture":
            return Self.rejected(commandID, code: "unsupported", message: "iOS 画中画尚未启用")
        case "set-frame-interpolation":
            return Self.rejected(commandID, code: "unsupported", message: "iOS 播放器暂不支持模型补帧")
        case "set-hdr":
            return Self.rejected(commandID, code: "unsupported", message: "iOS 播放器暂不支持 HDR 输出")
        case "previous-item": controller.previousEpisode()
        case "next-item": controller.nextEpisode()
        case "retry": controller.retry()
        case "close": closePlayer(animated: true)
        default:
            return Self.rejected(commandID, code: "unsupported", message: "当前播放器不支持该命令")
        }
        return Self.accepted(commandID)
    }

    /** 从受控来源创建播放请求并展示原生页面。 */
    private func launch(_ command: [String: Any], sessionID: String) throws {
        guard let source = command["source"] as? [String: Any] else {
            throw AniPlayerPluginError.invalidCommand("播放器加载命令缺少媒体来源")
        }
        let taskID = try Self.requiredString(source, key: "taskId")
        let title = try Self.requiredString(source, key: "title")
        let mediaURL = try Self.mediaURL(Self.requiredString(source, key: "uri"))
        let presentation = PlayerLaunchParser.parsePresentation(source, fallbackTitle: title)
        let fileIndex = source["fileIndex"] as? Int
        let episodeID = fileIndex.map { "\(taskID):\($0)" } ?? taskID
        let subtitles = (source["subtitles"] as? [[String: Any]] ?? []).enumerated().compactMap {
            index, value -> PlayerSubtitle? in
            guard
                let uri = value["uri"] as? String,
                let url = try? Self.mediaURL(uri)
            else { return nil }
            return PlayerSubtitle(
                id: (value["id"] as? String) ?? "subtitle-\(index)",
                label: (value["label"] as? String) ?? "字幕 \(index + 1)",
                url: url,
                language: value["language"] as? String,
                isDefault: (value["default"] as? Bool) ?? (index == 0)
            )
        }
        let durationMilliseconds = Self.milliseconds(source["durationSeconds"] as? Double)
        let episode = PlayerEpisode(
            id: episodeID,
            title: title,
            episodeLabel: title,
            mediaURL: mediaURL,
            durationMilliseconds: durationMilliseconds,
            subtitles: subtitles
        )
        let request = PlayerLaunchRequest(
            sessionID: sessionID,
            animeTitle: presentation.animeTitle,
            synopsis: presentation.synopsis,
            artworkURL: presentation.artworkURL,
            episodes: [episode],
            activeIndex: 0,
            startPositionMilliseconds: Self.milliseconds(command["startPositionSeconds"] as? Double),
            autoplay: true
        )
        let playerController = controller ?? MobileVLCPlayerController()
        controller = playerController
        playerController.initialize(request)
        activeSessionID = sessionID
        activeSource = source
        sequence = 0
        try presentPlayerIfNeeded(playerController)
        NSLog(
            "AniPlayerPlugin native player launched session=%@ artwork=%@",
            sessionID,
            presentation.artworkURL == nil ? "false" : "true"
        )
    }

    /** 在当前 Tauri iOS 窗口上全屏展示 SwiftUI 播放页。 */
    private func presentPlayerIfNeeded(_ controller: MobileVLCPlayerController) throws {
        if let hostingController, hostingController.presentingViewController != nil {
            return
        }
        guard let presenter = Self.topViewController() else {
            throw AniPlayerPluginError.unavailable("找不到可承载播放器的 iOS 窗口")
        }
        let host = UIHostingController(
            rootView: PlayerScreen(controller: controller) { [weak self] in
                self?.closePlayer(animated: true)
            }
        )
        host.modalPresentationStyle = .fullScreen
        host.isModalInPresentation = true
        hostingController = host
        presenter.present(host, animated: true) {
            host.presentationController?.delegate = self
        }
    }

    /** 停止媒体、关闭页面并清除真实路径引用。 */
    private func closePlayer(animated: Bool) {
        controller?.close()
        let host = hostingController
        hostingController = nil
        controller = nil
        activeSessionID = nil
        activeSource = nil
        sequence = 0
        if host?.presentingViewController != nil {
            host?.dismiss(animated: animated)
        }
    }

    /** 把 SwiftUI 状态归一为跨平台播放器快照。 */
    private func snapshotPayload() -> [String: Any]? {
        guard
            let controller,
            let sessionID = activeSessionID,
            let source = activeSource,
            controller.snapshot.sessionID == sessionID
        else { return nil }
        let snapshot = controller.snapshot
        sequence &+= 1
        let taskID = (source["taskId"] as? String) ?? ""
        let fileIndex = source["fileIndex"] as? Int
        let itemID = fileIndex.map { "\(taskID):\($0)" } ?? taskID
        var item: [String: Any] = [
            "id": itemID,
            "taskId": taskID,
            "title": (source["title"] as? String) ?? ""
        ]
        if let fileIndex { item["fileIndex"] = fileIndex }
        if let duration = source["durationSeconds"] as? Double { item["durationSeconds"] = duration }
        var payload: [String: Any] = [
            "sessionId": sessionID,
            "sequence": sequence,
            "backend": "libvlc",
            "platform": "ios",
            "status": Self.statusValue(snapshot.status),
            "capabilities": Self.capabilitiesPayload(),
            "source": source,
            "playlist": ["items": [item], "activeItemId": itemID],
            "positionSeconds": Double(snapshot.positionMilliseconds) / 1_000,
            "durationSeconds": Double(snapshot.durationMilliseconds) / 1_000,
            "bufferedSeconds": 0,
            "volume": Double(snapshot.volume) / 100,
            "muted": snapshot.muted,
            "playbackRate": Double(snapshot.playbackRate),
            "audioTracks": Self.trackPayload(snapshot.audioTracks, kind: "audio"),
            "subtitleTracks": Self.trackPayload(snapshot.subtitleTracks, kind: "subtitle"),
            "subtitleScale": snapshot.subtitleScale,
            "videoEnhancement": "off",
            "videoEnhancementDegraded": false,
            "frameInterpolation": "off",
            "hdr": "off",
            "enhancementDiagnostics": [
                "pipeline": "libvlc",
                "droppedFrames": 0,
                "hdrCapabilities": ["sourceHdr": false, "rendererHdr": false, "displayHdr": false]
            ],
            "aspectRatio": Self.aspectRatioValue(snapshot.aspectRatio),
            "fullscreen": Self.isLandscape,
            "pictureInPicture": false
        ]
        if let message = snapshot.errorMessage {
            payload["error"] = [
                "code": "decoder",
                "message": message,
                "recoverable": true,
                "recoveryActions": ["retry", "close"]
            ]
        }
        return payload
    }

    /** iOS 16+ 请求目标方向，iOS 15 保留系统手动旋转。 */
    private func setFullscreen(_ fullscreen: Bool) {
        guard let scene = Self.activeWindowScene else { return }
        if #available(iOS 16.0, *) {
            let preferences = UIWindowScene.GeometryPreferences.iOS(
                interfaceOrientations: fullscreen ? .landscape : .portrait
            )
            scene.requestGeometryUpdate(preferences) { error in
                NSLog("AniPlayerPlugin orientation update failed: %@", error.localizedDescription)
            }
        }
        UIViewController.attemptRotationToDeviceOrientation()
    }

    /** 读取非空字符串字段。 */
    private static func requiredString(_ object: [String: Any], key: String) throws -> String {
        guard let value = object[key] as? String, !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw AniPlayerPluginError.invalidCommand("播放器字段 \(key) 无效")
        }
        return value
    }

    /** 将网络 URI 或绝对沙盒路径转换为 URL。 */
    private static func mediaURL(_ value: String) throws -> URL {
        if let url = URL(string: value), url.scheme != nil { return url }
        guard value.hasPrefix("/") else {
            throw AniPlayerPluginError.invalidCommand("媒体路径不是受支持的绝对路径")
        }
        return URL(fileURLWithPath: value)
    }

    /** 将有限秒数安全换算为毫秒。 */
    private static func milliseconds(_ seconds: Double?) -> Int64 {
        guard let seconds, seconds.isFinite, seconds > 0 else { return 0 }
        return Int64(min(seconds * 1_000, Double(Int64.max)))
    }

    /** 将公共比例枚举映射到 iOS 离散比例。 */
    private static func aspectRatio(_ command: [String: Any]) -> PlayerAspectRatio? {
        let kind = (command["aspectRatio"] as? String) ?? "default"
        let value = command["value"] as? String
        switch kind == "custom" ? value : kind {
        case "default", "fit":
            return .automatic
        case "16:9":
            return .widescreen
        case "4:3":
            return .standard
        default:
            return nil
        }
    }

    /** 构造成功命令响应。 */
    private static func accepted(_ commandID: String) -> [String: Any] {
        ["commandId": commandID, "accepted": true]
    }

    /** 构造稳定错误码的拒绝响应。 */
    private static func rejected(_ commandID: String, code: String, message: String) -> [String: Any] {
        [
            "commandId": commandID,
            "accepted": false,
            "error": [
                "code": code,
                "message": message,
                "recoverable": code != "unsupported",
                "recoveryActions": ["close"]
            ]
        ]
    }

    /** 返回 iOS MobileVLCKit 的能力字典。 */
    private static func capabilitiesPayload() -> [String: Any] {
        [
            "backend": "libvlc",
            "platform": "ios",
            "availability": "available",
            "canSeek": true,
            "canSetVolume": true,
            "canMute": true,
            "playbackRates": [0.5, 0.75, 1.0, 1.25, 1.5, 2.0],
            "supportsAudioTracks": true,
            "supportsSubtitleTracks": true,
            "supportsSubtitleScale": true,
            "supportsVideoEnhancement": false,
            "supportsFrameInterpolation": false,
            "supportsModelEnhancement": false,
            "supportsAspectRatio": true,
            "supportsFullscreen": true,
            "supportsPictureInPicture": false,
            "supportsPlaylistNavigation": true,
            "supportsDirectPlayback": true,
            "supportsTranscodingFallback": false,
            "supportsHdr": false
        ]
    }

    /** 将平台轨道转换为公共轨道结构。 */
    private static func trackPayload(_ tracks: [PlayerTrack], kind: String) -> [[String: Any]] {
        tracks.map { track in
            [
                "id": String(track.id),
                "kind": kind,
                "label": track.label,
                "selected": track.selected
            ]
        }
    }

    /** 返回跨语言稳定的播放状态值。 */
    private static func statusValue(_ status: PlayerStatus) -> String {
        switch status {
        case .idle: "idle"
        case .loading: "loading"
        case .ready: "ready"
        case .playing: "playing"
        case .paused: "paused"
        case .buffering: "buffering"
        case .ended: "ended"
        case .error: "error"
        }
    }

    /** 返回跨语言稳定的画面比例值。 */
    private static func aspectRatioValue(_ aspectRatio: PlayerAspectRatio) -> String {
        switch aspectRatio {
        case .automatic: "default"
        case .widescreen: "16:9"
        case .standard: "4:3"
        }
    }

    /** 将桥接对象编码为 Rust 可解码的 JSON。 */
    private static func encode(_ object: Any) -> String {
        guard
            JSONSerialization.isValidJSONObject(object),
            let data = try? JSONSerialization.data(withJSONObject: object),
            let value = String(data: data, encoding: .utf8)
        else { return "{}" }
        return value
    }

    private static var activeWindowScene: UIWindowScene? {
        UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first { $0.activationState == .foregroundActive }
    }

    private static var isLandscape: Bool {
        activeWindowScene?.interfaceOrientation.isLandscape == true
    }

    /** 查找当前可展示全屏播放器的最上层控制器。 */
    private static func topViewController() -> UIViewController? {
        let root = activeWindowScene?.windows.first(where: \.isKeyWindow)?.rootViewController
        var current = root
        while let presented = current?.presentedViewController {
            current = presented
        }
        return current
    }
}

@_cdecl("init_plugin_ani_player")
/** 向 Tauri iOS 注册播放器插件实例。 */
func initPlugin() -> Plugin {
    AniPlayerPlugin()
}

File: crates/tauri-plugin-ani-player/android/src/main/java/dev/ani/tracker/player/AniPlayerPlugin.kt
package dev.ani.tracker.player

import android.app.Activity
import android.net.Uri
import android.util.Log
import android.webkit.WebView
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import dev.ani.tracker.android.player.NativePlayerBridge
import dev.ani.tracker.android.player.PlayerEpisode
import dev.ani.tracker.android.player.PlayerLaunchContract
import dev.ani.tracker.android.player.PlayerLaunchRequest
import dev.ani.tracker.android.player.PlayerSubtitle
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicLong

@InvokeArg
class DispatchArgs {
    lateinit var commandJson: String
}

/** 将 Rust PlayerTransport 连接到 Android libVLC Activity。 */
@TauriPlugin
class AniPlayerPlugin(private val activity: Activity) : Plugin(activity) {
    private val executor = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "ani-player-plugin")
    }
    private val sequence = AtomicLong(0L)
    @Volatile private var activeSessionId: String? = null
    @Volatile private var activeSource: JSONObject? = null

    /** 记录插件加载，不在 WebView 中注册原生文件命令。 */
    override fun load(webView: WebView) {
        Log.i(LOG_TAG, "Tauri Android player plugin loaded")
    }

    /** 返回 Android 内置 libVLC 的稳定能力集合。 */
    @Command
    fun capabilities(invoke: Invoke) {
        invoke.resolve(JSObject().put("capabilitiesJson", capabilitiesJson().toString()))
    }

    /** 执行 Rust 已校验并解析真实路径的播放器命令。 */
    @Command
    fun dispatch(invoke: Invoke) {
        val args = try {
            invoke.parseArgs(DispatchArgs::class.java)
        } catch (error: Exception) {
            invoke.reject("播放器命令参数无效", "invalid_player_command", error)
            return
        }
        executor.execute {
            try {
                val command = JSONObject(args.commandJson)
                val result = if (command.optString("type") == "load") {
                    launch(command)
                } else {
                    dispatchToActivity(command)
                }
                invoke.resolve(JSObject().put("resultJson", result.toString()))
            } catch (error: Exception) {
                Log.e(LOG_TAG, "Android player command failed", error)
                invoke.reject("Android 播放器命令失败", "player_command_failed", error)
            }
        }
    }

    /** 返回当前 Activity 快照；尚未打开播放器时返回空。 */
    @Command
    fun snapshot(invoke: Invoke) {
        executor.execute {
            try {
                val raw = NativePlayerBridge.snapshot().getOrThrow()
                val snapshot = raw?.let(::buildSnapshot)
                invoke.resolve(JSObject().put("snapshotJson", snapshot?.toString()))
            } catch (error: Exception) {
                Log.e(LOG_TAG, "Android player snapshot failed", error)
                invoke.reject("Android 播放器状态读取失败", "player_snapshot_failed", error)
            }
        }
    }

    /** 幂等关闭 Activity 并清除原生会话引用。 */
    @Command
    fun shutdown(invoke: Invoke) {
        executor.execute {
            try {
                NativePlayerBridge.shutdown().getOrThrow()
                activeSessionId = null
                activeSource = null
                invoke.resolve(JSObject().put("stopped", true))
            } catch (error: Exception) {
                Log.e(LOG_TAG, "Android player shutdown failed", error)
                invoke.reject("Android 播放器关闭失败", "player_shutdown_failed", error)
            }
        }
    }

    /** 插件销毁时释放串行执行器，Activity 生命周期由系统管理。 */
    override fun onDestroy() {
        executor.shutdownNow()
    }

    /** 从受控来源创建 Intent 并启动原生播放器。 */
    private fun launch(command: JSONObject): JSONObject {
        val commandId = command.getString("commandId")
        val sessionId = command.getString("sessionId")
        val source = command.getJSONObject("source")
        val mediaUri = normalizeUri(source.getString("uri"))
        val subtitles = source.optJSONArray("subtitles").toObjectList().mapIndexed { index, subtitle ->
            PlayerSubtitle(
                id = subtitle.optString("id", "subtitle-$index"),
                label = subtitle.optString("label", "字幕 ${index + 1}"),
                uri = normalizeUri(subtitle.getString("uri")),
                language = subtitle.optString("language").takeIf(String::isNotBlank),
                isDefault = subtitle.optBoolean("default", index == 0)
            )
        }
        val fileIndex = source.optInt("fileIndex", -1).takeIf { it >= 0 }
        val episodeId = buildString {
            append(source.getString("taskId"))
            fileIndex?.let { append(":").append(it) }
        }
        val request = PlayerLaunchRequest(
            sessionId = sessionId,
            animeTitle = source.optString("animeTitle").takeIf(String::isNotBlank)
                ?: source.getString("title"),
            description = source.optString("description"),
            artworkUri = source.optString("artworkUri").takeIf(String::isNotBlank),
            episodes = listOf(
                PlayerEpisode(
                    id = episodeId,
                    title = source.getString("title"),
                    episodeLabel = source.getString("title"),
                    uri = mediaUri,
                    durationMillis = (source.optDouble("durationSeconds", 0.0) * 1_000.0).toLong(),
                    subtitles = subtitles
                )
            ),
            activeIndex = 0,
            startPositionMillis = (command.optDouble("startPositionSeconds", 0.0) * 1_000.0).toLong(),
            autoplay = true
        )
        activity.runOnUiThread {
            activity.startActivity(PlayerLaunchContract.createIntent(activity, request))
        }
        activeSessionId = sessionId
        activeSource = JSONObject(source.toString())
        sequence.set(0L)
        Log.i(LOG_TAG, "Android native player launched session=$sessionId, artwork=${request.artworkUri != null}")
        return accepted(commandId)
    }

    /** 将非加载动作转发到当前 PlayerActivity。 */
    private fun dispatchToActivity(command: JSONObject): JSONObject {
        val commandId = command.getString("commandId")
        if (activeSessionId != command.optString("sessionId")) {
            return rejected(commandId, "resource-unavailable", "播放器会话已切换")
        }
        if (command.optString("type") == "set-picture-in-picture") {
            return rejected(commandId, "unsupported", "Android 画中画尚未启用")
        }
        if (command.optString("type") == "set-frame-interpolation") {
            return rejected(commandId, "unsupported", "Android 播放器暂不支持模型补帧")
        }
        if (command.optString("type") == "set-hdr") {
            return rejected(commandId, "unsupported", "Android 播放器暂不支持 HDR 输出")
        }
        val accepted = NativePlayerBridge.dispatch(command).getOrThrow()
        if (!accepted) return rejected(commandId, "unsupported", "当前播放器不支持该命令")
        if (command.optString("type") == "close") {
            activeSessionId = null
            activeSource = null
        }
        return accepted(commandId)
    }

    /** 把 Activity 状态归一为跨平台播放器快照。 */
    private fun buildSnapshot(raw: JSONObject): JSONObject? {
        val sessionId = activeSessionId ?: return null
        if (raw.optString("sessionId") != sessionId) return null
        val source = activeSource ?: return null
        val errorMessage = raw.optString("errorMessage").takeIf(String::isNotBlank)
        val sourceCopy = JSONObject(source.toString())
        val itemId = buildString {
            append(sourceCopy.optString("taskId"))
            if (sourceCopy.has("fileIndex")) append(":").append(sourceCopy.optInt("fileIndex"))
        }
        return JSONObject().apply {
            put("sessionId", sessionId)
            put("sequence", sequence.incrementAndGet())
            put("backend", "libvlc")
            put("platform", "android")
            put("status", raw.optString("status", "idle"))
            put("capabilities", capabilitiesJson())
            put("source", sourceCopy)
            put("playlist", JSONObject().apply {
                put("items", JSONArray().put(JSONObject().apply {
                    put("id", itemId)
                    put("taskId", sourceCopy.optString("taskId"))
                    if (sourceCopy.has("fileIndex")) put("fileIndex", sourceCopy.optInt("fileIndex"))
                    put("title", sourceCopy.optString("title"))
                    if (sourceCopy.has("durationSeconds")) {
                        put("durationSeconds", sourceCopy.optDouble("durationSeconds"))
                    }
                }))
                put("activeItemId", itemId)
            })
            put("positionSeconds", raw.optDouble("positionSeconds", 0.0))
            put("durationSeconds", raw.optDouble("durationSeconds", 0.0))
            put("bufferedSeconds", 0.0)
            put("volume", raw.optDouble("volume", 0.7))
            put("muted", raw.optBoolean("muted"))
            put("playbackRate", raw.optDouble("playbackRate", 1.0))
            put("audioTracks", normalizeTracks(raw.optJSONArray("audioTracks"), "audio"))
            put("subtitleTracks", normalizeTracks(raw.optJSONArray("subtitleTracks"), "subtitle"))
            put("subtitleScale", raw.optInt("subtitleScale", 100))
            put("videoEnhancement", "off")
            put("videoEnhancementDegraded", false)
            put("frameInterpolation", "off")
            put("hdr", "off")
            put("enhancementDiagnostics", JSONObject().apply {
                put("pipeline", "libvlc")
                put("droppedFrames", 0)
                put("hdrCapabilities", JSONObject().apply {
                    put("sourceHdr", false)
                    put("rendererHdr", false)
                    put("displayHdr", false)
                })
            })
            val aspectRatio = raw.opt("aspectRatio")
                ?.takeUnless { it == JSONObject.NULL }
                ?.toString()
                ?.takeIf(String::isNotBlank)
                ?: "default"
            put("aspectRatio", aspectRatio)
            put("fullscreen", raw.optBoolean("fullscreen"))
            put("pictureInPicture", false)
            errorMessage?.let {
                put("error", JSONObject().apply {
                    put("code", "decoder")
                    put("message", it)
                    put("recoverable", true)
                    put("recoveryActions", JSONArray().put("retry").put("close"))
                })
            }
        }
    }

    /** 返回 Android libVLC 的能力字典。 */
    private fun capabilitiesJson(): JSONObject = JSONObject().apply {
        put("backend", "libvlc")
        put("platform", "android")
        put("availability", "available")
        put("canSeek", true)
        put("canSetVolume", true)
        put("canMute", true)
        put("playbackRates", JSONArray(listOf(0.5, 0.75, 1.0, 1.25, 1.5, 2.0)))
        put("supportsAudioTracks", true)
        put("supportsSubtitleTracks", true)
        put("supportsSubtitleScale", true)
        put("supportsVideoEnhancement", false)
        put("supportsFrameInterpolation", false)
        put("supportsModelEnhancement", false)
        put("supportsAspectRatio", true)
        put("supportsFullscreen", true)
        put("supportsPictureInPicture", false)
        put("supportsPlaylistNavigation", true)
        put("supportsDirectPlayback", true)
        put("supportsTranscodingFallback", false)
        put("supportsHdr", false)
    }

    /** 将平台轨道转换为公共轨道结构。 */
    private fun normalizeTracks(tracks: JSONArray?, kind: String): JSONArray = JSONArray().apply {
        tracks.toObjectList().forEach { track ->
            put(JSONObject().apply {
                put("id", track.optString("id"))
                put("kind", kind)
                put("label", track.optString("label"))
                put("selected", track.optBoolean("selected"))
            })
        }
    }

    /** 将绝对文件路径转换为 libVLC 可识别的 file URI。 */
    private fun normalizeUri(value: String): String {
        val parsed = Uri.parse(value)
        return if (parsed.scheme.isNullOrBlank()) Uri.fromFile(File(value)).toString() else value
    }

    /** 构造成功命令响应。 */
    private fun accepted(commandId: String): JSONObject = JSONObject().apply {
        put("commandId", commandId)
        put("accepted", true)
    }

    /** 构造稳定错误码的拒绝响应。 */
    private fun rejected(commandId: String, code: String, message: String): JSONObject = JSONObject().apply {
        put("commandId", commandId)
        put("accepted", false)
        put("error", JSONObject().apply {
            put("code", code)
            put("message", message)
            put("recoverable", code != "unsupported")
            put("recoveryActions", JSONArray().put("close"))
        })
    }

    /** 将可空 JSON 数组安全转换为对象列表。 */
    private fun JSONArray?.toObjectList(): List<JSONObject> {
        if (this == null) return emptyList()
        return buildList {
            for (index in 0 until length()) optJSONObject(index)?.let(::add)
        }
    }

    companion object {
        private const val LOG_TAG = "AniPlayerPlugin"
    }
}
